Clean up model editing functions.

luciansmith opened the pull request:
This allows the user to call 'regenerate' directly, instead of having to call it by proxy when adding/removing SBML elements.  It also cleans up a lot of the model editing functions:
  * Only does some validation when 'forceRegenerate' is on, on the assumption that until the user wants to regenerate the model, it might be in an inconsistent state.
  * Calls 'getReactant'/'getProduct' directly instead of 'getReactantList', etc.
  * Consolidates checks when adding assignment and rate rules.
  * Changes the constant/boundarySpecies attributes as appropriate (might want to throw instead?)
  * Collects a set of variables to check, instead of a vector (to eliminate duplicate checking).
  * Properly cleans up after species references with rules when deleting reactions.

Finally, fixes tests that had added a rate rules/assignment rules for non-boundary species:  they're now properly boundary species when constructed.

luciansmith committed:
Clean up model editing functions.

This allows the user to call 'regenerate' directly, instead of having to call it by proxy when adding/removing SBML elements.  It also cleans up a lot of the model editing functions:
  * Only does some validation when 'forceRegenerate' is on, on the assumption that until the user wants to regenerate the model, it might be in an inconsistent state.
  * Calls 'getReactant'/'getProduct' directly instead of 'getReactantList', etc.
  * Consolidates checks when adding assignment and rate rules.
  * Changes the constant/boundarySpecies attributes as appropriate (might want to throw instead?)
  * Collects a set of variables to check, instead of a vector (to eliminate duplicate checking).
  * Properly cleans up after species references with rules when deleting reactions.

Finally, fixes tests that had added a rate rules/assignment rules for non-boundary species:  they're now properly boundary species when constructed.

luciansmith commented:
Just to make sure these changes would indeed be helpful.  (We also need to decide whether to fix or throw when there's a rule added for a non-constant element or non-boundary species.  Right now, I fix, but maybe a throw would be more obvious to the user.)

CiaranWelsh commented:
> Just to make sure these changes would indeed be helpful. (We also need to decide whether to fix or throw when there's a rule  added for a non-constant element or non-boundary species. Right now, I fix, but maybe a throw would be more obvious to the user.)

Throw with a helpful error message is more likely to provide the user with the know how for future so might be the better option. However, we can probably achieve the same effect with a warning and also have the convenience of an automatic fix. 

I've copied over your changes.

```
#include "rr/rrRoadRunner.h"
using namespace rr;

int main(){
    RoadRunner rr;

    rr.addCompartment("Comp", 1, false);
    rr.addSpecies("S1", "Comp", 10.0, false, false, "", false);
    rr.addSpecies("S2", "Comp", 0.0, false, false, "", false);
    rr.addParameter("k1", 0.1, false);
    rr.addReaction("R1", std::vector<std::string>({"S1"}), std::vector<std::string>({"S2"}), "k1*S1", false);

    rr.regenerate(true, false);

    SimulateOptions options;
    options.start = 0;
    options.steps = 101;
    options.duration = 100;

}
```
This is certainly an improvement since it means I no longer have to use a copy of rr that has diverged from the main code base. My only suggestion is to set a default parameter on the first argument to force regenerate to true. i.e. I'd like to call it using `rr.regenerate()` rather than `rr.regenerate(true)`, a small thing. 

Slight divergence on topic - I had to copy across your sources to test this out, rather than pull because my build system is different to the main repo and itd be a large merge job. If you let me know when might be a good time to try merging, I'll try it out on a fresh branch. By "good time", I mean once your at a good place with other merges, fixing tests etc.

luciansmith committed:
Throw if trying to add a rule for a floating species.

Also had to fix a bunch of tests that accidentally created invalid models.  We chose to throw for boundary conditions but not for constant because the user cannot actually set 'constant' when creating any element, but must do it separately if they don't want the default.
(Also fixed the wording in our other throws, since they can be called both for setting assignment rules and rate rules.)

luciansmith committed:
Fix other tests, too.

luciansmith committed:
Give 'regenerate' default arguments.

luciansmith commented:
OK!  I went with a throw for trying to add a rule for a floating species, since the user can set that in the species construction, but silently change anything else to constant=false behind the scenes, since the user can't set that during construction, but only afterwards.  Also gave 'regenerate' a default value.
